drivers/rtc/rtc_rv3028.c

```diff
@@ -0,0 +1,980 @@
+#define RV3028_CLKOUT_PORIE             BIT(3)
+#define RV3028_CLKOUT_CLKSY             BIT(6)
+#define RV3028_CLKOUT_CLKOE             BIT(7)
+
+#define RV3028_CLKOUT_FD_LOW            0x7
+
+#define RV3028_BACKUP_TCE               BIT(5)
+#define RV3028_BACKUP_TCR               GENMASK(1, 0)
+#define RV3028_BACKUP_BSM               GENMASK(3, 2)
+
+#define RV3028_BSM_LEVEL                0x3
+#define RV3028_BSM_DIRECT               0x1
+#define RV3028_BSM_DISABLED             0x0
+
+/* RV3028 EE command register values */
+#define RV3028_EEPROM_CMD_INIT          0x00
+#define RV3028_EEPROM_CMD_UPDATE        0x11
+#define RV3028_EEPROM_CMD_REFRESH       0x12
+#define RV3028_EEPROM_CMD_WRITE         0x21
+#define RV3028_EEPROM_CMD_READ          0x22
+
+#define RV3028_SECONDS_MASK             GENMASK(6, 0)
+#define RV3028_MINUTES_MASK             GENMASK(6, 0)
+#define RV3028_HOURS_AMPM               BIT(5)
+#define RV3028_HOURS_12H_MASK           GENMASK(4, 0)
+#define RV3028_HOURS_24H_MASK           GENMASK(5, 0)
+#define RV3028_DATE_MASK                GENMASK(5, 0)
+#define RV3028_WEEKDAY_MASK             GENMASK(2, 0)
+#define RV3028_MONTH_MASK               GENMASK(4, 0)
+#define RV3028_YEAR_MASK                GENMASK(7, 0)
+
+#define RV3028_ALARM_MINUTES_AE_M       BIT(7)
+#define RV3028_ALARM_MINUTES_MASK       GENMASK(6, 0)
+#define RV3028_ALARM_HOURS_AE_H         BIT(7)
+#define RV3028_ALARM_HOURS_AMPM         BIT(5)
+#define RV3028_ALARM_HOURS_12H_MASK     GENMASK(4, 0)
+#define RV3028_ALARM_HOURS_24H_MASK     GENMASK(5, 0)
+#define RV3028_ALARM_DATE_AE_WD         BIT(7)
+#define RV3028_ALARM_DATE_MASK          GENMASK(5, 0)
+
+/* The RV3028 only supports two-digit years. Leap years are correctly handled from 2000 to 2099 */
+#define RV3028_YEAR_OFFSET              (2000 - 1900)
+
+/* The RV3028 enumerates months 1 to 12 */
+#define RV3028_MONTH_OFFSET 1
+
+#define RV3028_EEBUSY_POLL_US           10000
+#define RV3028_EEBUSY_TIMEOUT_MS        100
+
+/* RTC alarm time fields supported by the RV3028 */
+#define RV3028_RTC_ALARM_TIME_MASK                                                                 \
+	(RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MONTHDAY)
+
+/* RTC time fields supported by the RV3028 */
+#define RV3028_RTC_TIME_MASK                                                                       \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_YEAR)
+
+/* Helper macro to guard int-gpios related code */
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios) && \
+	(defined(CONFIG_RTC_ALARM) || defined(CONFIG_RTC_UPDATE))
+#define RV3028_INT_GPIOS_IN_USE 1
+#endif
+
+struct rv3028_config {
+	const struct i2c_dt_spec i2c;
+#ifdef RV3028_INT_GPIOS_IN_USE
+	struct gpio_dt_spec gpio_int;
+#endif /* RV3028_INT_GPIOS_IN_USE */
+	uint8_t cof;
+	uint8_t backup;
+};
+
+struct rv3028_data {
+	struct k_mutex lock;
```

Comment:
Please use a `k_sem` for locking if the owner of the lock is not used (like it would be for a condvar or recursive locking)

Like this:
https://github.com/zephyrproject-rtos/zephyr/blob/593dfe13856df4d2919874037cbb5691b9f00a15/drivers/gnss/gnss_emul.c#L49-L61
and
https://github.com/zephyrproject-rtos/zephyr/blob/593dfe13856df4d2919874037cbb5691b9f00a15/drivers/gnss/gnss_emul.c#L470

Reply:
In terms of semantics, a mutex seemed like a more intuitive choice, but I honestly had no idea it was more expensive than semaphores 😊